packages/unlock-js/src/walletService.ts

```diff
@@ -997,4 +1013,150 @@ export default class WalletService extends UnlockService {
       callback
     )
   }
+
+  async swapAndCall({
+    params: {
+      network,
+      lockAddress,
+      srcTokenAddress,
+      amountInMax,
+      uniswapRouter,
+      callData,
+      swapCallData,
+    },
+  }: {
+    params: {
+      lockAddress: string
+      network: number
+      srcTokenAddress: string
+      amountInMax: string
+      uniswapRouter: string
+      swapCallData: string
+      callData: string
+    }
+  }) {
+    const networkConfig = this.networks[network]
+    const UnlockSwapPurchaser = networkConfig?.swapPurchaser
+    if (!UnlockSwapPurchaser) {
+      throw new Error('SwapPurchaser not available for this network')
+    }
+    const provider = this.providerForNetwork(network)
+    const swapPurchaserContract = new ethers.Contract(
+      UnlockSwapPurchaser,
+      UnlockSwapPurchaserABI,
+      provider
+    )
+    const swapPurchaser = swapPurchaserContract.connect(this.signer)
+
+    const tx = await swapPurchaser.swapAndCall(
+      lockAddress,
+      srcTokenAddress,
+      amountInMax,
+      uniswapRouter,
+      swapCallData,
+      callData,
+      {
+        value: amountInMax,
+      }
+    )
+    return tx
+  }
+
+  async swapAndPurchase({
```

Comment:
A lot of the `purchase` logic is reused here from the `purchase` function. Is there a way we could DRY things up a bit? I worry about diverging implementation when it is really doing mostly the same thing, no?